transportable-udfs-annotation-processor/src/main/java/com/linkedin/transport/processor/TransportProcessor.java

```diff
@@ -0,0 +1,189 @@
+  private Elements _elements;
+  private TypeMirror _topLevelStdUDFInterfaceType;
+  private TypeMirror _stdUDFClassType;
+  private UDFProperties _udfProperties;
+
+  @Override
+  public SourceVersion getSupportedSourceVersion() {
+    return SourceVersion.latestSupported();
+  }
+
+  @Override
+  public synchronized void init(ProcessingEnvironment processingEnv) {
+    super.init(processingEnv);
+    _types = processingEnv.getTypeUtils();
+    _elements = processingEnv.getElementUtils();
+    _topLevelStdUDFInterfaceType = _elements.getTypeElement(TopLevelStdUDF.class.getName()).asType();
+    _stdUDFClassType = _elements.getTypeElement(StdUDF.class.getName()).asType();
+    _udfProperties = new UDFProperties();
+  }
+
+  @Override
+  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+    try {
+      processImpl(roundEnv);
+    } catch (Exception e) {
+      // We don't allow exceptions of any kind to propagate to the compiler
+      try (StringWriter stringWriter = new StringWriter(); PrintWriter printWriter = new PrintWriter(stringWriter)) {
+        e.printStackTrace(printWriter);
+        fatalError(stringWriter.toString());
+      } catch (IOException ioe) {
+        fatalError("Could not close resources " + ioe);
+      }
+    }
+    // Universal processors should return false since other processor can be potentially acting on the same element
+    return false;
+  }
+
+  private void processImpl(RoundEnvironment roundEnv) {
+    if (roundEnv.processingOver()) {
+      generateUDFPropertiesFile();
+    } else {
+      processElements(roundEnv.getRootElements());
+    }
+  }
+
+  private void processElements(Set<? extends Element> elements) {
+    for (Element element : elements) {
+      // Check if the element is a non-abstract class which extends the StdUDF class
+      if (element.getKind().equals(ElementKind.CLASS) && !element.getModifiers().contains(Modifier.ABSTRACT)
+          && _types.isAssignable(element.asType(), _stdUDFClassType)) {
+        processUDFClass((TypeElement) element);
+      }
+    }
+  }
+
+  /**
+   * Finds the {@link TopLevelStdUDF} for the given {@link StdUDF} class and adds it to the list of discovered UDFs
+   */
+  private void processUDFClass(TypeElement udfClassElement) {
+    debug(String.format("Processing UDF Class: %s", udfClassElement.getQualifiedName()));
+
+    TypeElement closestTopLevelStdUDFInterface = getClosestTopLevelStdUDFInterface(udfClassElement);
+    if (closestTopLevelStdUDFInterface != null) {
+      _udfProperties.addUDF(closestTopLevelStdUDFInterface.getQualifiedName().toString(),
+          udfClassElement.getQualifiedName().toString());
+    }
+  }
+
+  /**
+   * Returns a {@link TypeElement} for the closest (in the type hierarchy) class/interface which implements the
+   * {@link TopLevelStdUDF} interface for a given {@link StdUDF} class
+   */
+  private TypeElement getClosestTopLevelStdUDFInterface(TypeElement udfClassElement) {
```

Comment:
I think we should define a set of accepted configurations, especially that some may not make sense, and the algorithm in this method should reflect that. Some example of complex configurations that we may think whether they have value or not:
udf1 extends udf2 (which implements topleveludf) implements topleveludf
udf1 extends udf2 (which implements iface1 which extends topleveludf) implements topleveludf 
udf1 extends udf2 (which implements iface1 which extends topleveludf) implements iface2 (which extends topleveludf) 
udf1 extends stdudf1 implements iface1 (which extends topleveludf), iface2 (which extends topleveludf)

For example, while Java can accept the last configuration by forcing the reimplementation of topleveludf inside udf1, I do not think from a UDF structure point of view, we should allow this configuration, because implementing iface1 and iface2 is not necessarily in the first place and does not seem to add value.

Reply:
Based on our discussion earlier this week, I came up with 4 rules which make these checks more strict

Rule 1: A UDF class implementing TopLevelStdUDF through a superclass is not supported as of now since it is hard to determine which interface we want to use as the base for the group of UDF overloading. This is not a common use case anyway. If required, use composition instead of inheritance.
Rule 2: More than one interface implementing TopLevelStdUDF can lead to ambiguity with respect to the name of the UDF and hence is not allowed.
Rule 3: A UDF class should implement TopLevelStdUDF either directly or indirectly through an interface only.
Rule 4: If the class does not directly implement TopLevelStdUDF, then verify that interface methods are not overriden since it can lead to UDF name ambiguity.

Reply:
As discussed yesterday, we will be checking if there is only one overriding of TopLevelStdUDF methods regardless of leaf/non-leaf, abstract/non-abstract.